src/lib/ThumbnailsSidebar.js

```diff
@@ -0,0 +1,175 @@
+import VirtualScroller from './VirtualScroller';
+
+const DEFAULT_THUMBNAILS_SIDEBAR_WIDTH = 150;
+const THUMBNAIL_WIDTH_MAX = 210;
+const THUMBNAIL_MARGIN = 15;
+
+class ThumbnailsSidebar {
+    /** @property {HTMLElement} - The anchor element for this ThumbnailsSidebar */
+    anchorEl;
+
+    /** @property {PDfViewer} - The PDFJS viewer instance */
+    pdfViewer;
+
+    /** @property {Object} - Cache for the thumbnail image elements */
+    thumbnailImageCache;
+
+    /**
+     * [constructor]
+     *
+     * @param {HTMLElement} element - the HTMLElement that will anchor the thumbnail sidebar
+     * @param {PDFViewer} pdfViewer - the PDFJS viewer
+     */
+    constructor(element, pdfViewer) {
+        this.anchorEl = element;
+        this.pdfViewer = pdfViewer;
+        this.thumbnailImageCache = {};
+
+        this.renderBasicThumbnail = this.renderBasicThumbnail.bind(this);
+        this.requestThumbnailImage = this.requestThumbnailImage.bind(this);
+        this.makeThumbnailImage = this.makeThumbnailImage.bind(this);
+        this.renderThumbnailImages = this.renderThumbnailImages.bind(this);
+    }
+
+    /**
+     * Destroys the thumbnails sidebar
+     *
+     * @return {void}
+     */
+    destroy() {
+        if (this.virtualScroller) {
+            this.virtualScroller.destroy();
+        }
+
+        this.thumbnailImageCache = null;
+        this.virtualScroller = null;
+        this.pdfViewer = null;
+    }
+
+    /**
+     * Initializes the Thumbnails Sidebar
+     *
+     * @return {void}
+     */
+    init() {
+        this.virtualScroller = new VirtualScroller(this.anchorEl);
+
+        // Get the first page of the document, and use its dimensions
+        // to set the thumbnails size of the thumbnails sidebar
+        this.pdfViewer.pdfDocument.getPage(1).then((page) => {
+            const desiredWidth = DEFAULT_THUMBNAILS_SIDEBAR_WIDTH;
+            const viewport = page.getViewport(1);
+            this.scale = desiredWidth / viewport.width;
+            this.pageRatio = viewport.width / viewport.height;
+            const scaledViewport = page.getViewport(this.scale);
```

Comment:
Can you leave a brief comment explaining how these different ratios will be used?